devtools/gradle/src/main/java/io/quarkus/gradle/QuarkusPluginExtension.java

```diff
@@ -55,6 +58,23 @@ void beforeTest(Test task) {
             final Path serializedModel = QuarkusGradleUtils.serializeAppModel(appModel, task);
             props.put(BootstrapConstants.SERIALIZED_APP_MODEL, serializedModel.toString());
 
+            // Identify the folder containing the sources associated with this test task
+            StringJoiner joiner = new StringJoiner(",");
+            getSourceSets().stream()
+                    .filter(sourceSet -> Objects.equals(
+                            task.getTestClassesDirs().getAsPath(),
+                            sourceSet.getOutput().getClassesDirs().getAsPath()))
+                    .flatMap(sourceSet -> sourceSet.getOutput().getClassesDirs().getFiles().stream())
+                    .filter(File::exists)
+                    .distinct()
+                    .forEach(testSrcDir -> {
+                        String mapping = String.format("%s:%s",
+                                project.relativePath(testSrcDir),
+                                project.relativePath(outputDirectory()));
+                        joiner.add(mapping);
+                    });
+            task.environment(BootstrapConstants.TEST_TO_MAIN_MAPPINGS, joiner.toString());
```

Comment:
Instead of using `forEach` and an external `StringJoiner`, would it be better to use something more functional?

```suggestion
            String fileList = getSourceSets().stream()
                    .filter(sourceSet -> Objects.equals(
                            task.getTestClassesDirs().getAsPath(),
                            sourceSet.getOutput().getClassesDirs().getAsPath()))
                    .flatMap(sourceSet -> sourceSet.getOutput().getClassesDirs().getFiles().stream())
                    .filter(File::exists)
                    .distinct()
                    .map(testSrcDir -> String.format("%s:%s",
                                project.relativePath(testSrcDir),
                                project.relativePath(outputDirectory())))
                    .collect( Collectors.joining( "," ) );
            task.environment(BootstrapConstants.TEST_TO_MAIN_MAPPINGS, fileList);
```